block-storage/src/main/scala/coop/rchain/blockstorage/LMDBBlockStore.scala

```diff
@@ -0,0 +1,133 @@
+package coop.rchain.blockstorage
+
+import java.nio.ByteBuffer
+import java.nio.file.Path
+
+import scala.collection.JavaConverters._
+import scala.language.higherKinds
+
+import cats._
+import cats.effect.{ExitCase, Sync}
+import cats.implicits._
+import com.google.protobuf.ByteString
+import coop.rchain.blockstorage.BlockStore.BlockHash
+import coop.rchain.casper.protocol.BlockMessage
+import coop.rchain.metrics.Metrics
+import org.lmdbjava._
+import org.lmdbjava.DbiFlags.MDB_CREATE
+import scala.util.control.NonFatal
+
+class LMDBBlockStore[F[_]] private (val env: Env[ByteBuffer], path: Path, blocks: Dbi[ByteBuffer])(
+    implicit
+    syncF: Sync[F],
+    metricsF: Metrics[F])
+    extends BlockStore[F] {
+
+  import LMDBBlockStore.MetricNamePrefix
+
+  implicit class RichBlockHash(byteVector: BlockHash) {
+
+    def toDirectByteBuffer: ByteBuffer = {
+      val buffer: ByteBuffer = ByteBuffer.allocateDirect(byteVector.size)
+      byteVector.copyTo(buffer)
+      // TODO: get rid of this:
+      buffer.flip()
+      buffer
+    }
+  }
+
+  def put(f: => (BlockHash, BlockMessage)): F[Unit] =
+    for {
+      _ <- metricsF.incrementCounter(MetricNamePrefix + "put")
+      ret <- syncF.bracket(syncF.delay(env.txnWrite())) { txn =>
+              syncF.delay {
+                val (blockHash, blockMessage) = f
+                blocks.put(txn,
+                           blockHash.toDirectByteBuffer,
+                           blockMessage.toByteString.toDirectByteBuffer)
+                txn.commit()
```

Comment:
I am looking at `Dbi.put` method
https://github.com/lmdbjava/lmdbjava/blob/6a3f8acfdbd9fac946e53dac41d01997646c5cf2/src/main/java/org/lmdbjava/Dbi.java#L386-L391

and it looks like it already takes care of creating a transaction and `commit`-ing.

Reply:
But you're doing that for the `put` as well.